rust/lance/src/dataset.rs

```diff
@@ -4675,4 +4682,242 @@ mod tests {
+                    field_b.clone(),
+                    Arc::new(Int32Array::from(vec![1])) as ArrayRef,
+                ),
+                (field_a.clone(), Arc::new(Int32Array::from(vec![2]))),
+            ]))],
+        )
+        .unwrap();
+        let reader = RecordBatchIterator::new(vec![Ok(batch)], just_b_a.clone());
+        let dataset = Dataset::write(reader, test_uri, Some(append_options.clone()))
+            .await
+            .unwrap();
+        let fragments = dataset.get_fragments();
+        assert_eq!(fragments.len(), 1);
+        assert_eq!(fragments[0].metadata.files.len(), 1);
+        assert_eq!(&fragments[0].metadata.files[0].fields, &[0, 2, 1]);
+        assert_eq!(&fragments[0].metadata.files[0].column_indices, &[0, 1, 2]);
+
+        // Can insert c, b
+        let just_c_b = Arc::new(ArrowSchema::new(vec![ArrowField::new(
+            "s",
+            DataType::Struct(vec![field_c.clone(), field_b.clone()].into()),
+            true,
+        )]));
+        let batch = RecordBatch::try_new(
+            just_c_b.clone(),
+            vec![Arc::new(StructArray::from(vec![
+                (
+                    field_c.clone(),
+                    Arc::new(Int32Array::from(vec![4])) as ArrayRef,
+                ),
+                (field_b.clone(), Arc::new(Int32Array::from(vec![3]))),
+            ]))],
+        )
+        .unwrap();
+        let reader = RecordBatchIterator::new(vec![Ok(batch)], just_c_b.clone());
+        let dataset = Dataset::write(reader, test_uri, Some(append_options.clone()))
+            .await
+            .unwrap();
+        let fragments = dataset.get_fragments();
+        assert_eq!(fragments.len(), 2);
+        assert_eq!(fragments[1].metadata.files.len(), 1);
+        assert_eq!(&fragments[1].metadata.files[0].fields, &[0, 3, 2]);
+        assert_eq!(&fragments[1].metadata.files[0].column_indices, &[0, 1, 2]);
+
+        // Can't insert a, c (b is non-nullable)
+        let just_a_c = Arc::new(ArrowSchema::new(vec![ArrowField::new(
+            "s",
+            DataType::Struct(vec![field_a.clone(), field_c.clone()].into()),
+            true,
+        )]));
+        let batch = RecordBatch::try_new(
+            just_a_c.clone(),
+            vec![Arc::new(StructArray::from(vec![
+                (
+                    field_a.clone(),
+                    Arc::new(Int32Array::from(vec![5])) as ArrayRef,
+                ),
+                (field_c.clone(), Arc::new(Int32Array::from(vec![6]))),
+            ]))],
+        )
+        .unwrap();
+        let reader = RecordBatchIterator::new(vec![Ok(batch)], just_a_c.clone());
+        let res = Dataset::write(reader, test_uri, Some(append_options)).await;
+        assert!(
+            matches!(res, Err(Error::SchemaMismatch { .. })),
+            "Expected Error::SchemaMismatch, got {:?}",
+            res
+        );
+
+        // Can scan and get all data
+        let data = dataset.scan().try_into_batch().await.unwrap();
```

Comment:
This fails with

```
IO { source: "External error: Invalid user input: there were more fields in the schema than provided column indices / infos, rust/lance-encoding/src/decoder.rs:427:17", location: Location { file: "rust/lance/src/dataset/scanner.rs", line: 2084, column: 83 } }
```

I think it has to do with these line, but I'm not quite sure what the solution is here:

https://github.com/lancedb/lance/blob/ceaf49c9d24eab6cf2dd61712849cb4f3bd14d69/rust/lance-file/src/v2/reader.rs#L701-L719